docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,66 @@
+# Title: Password Key Derivation
+
+## Abstract
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
```

Comment:
I can see a rationale for having the _possibility_ of bucket encryption keys that aren't derived from the root, but I disagree pretty strongly that that should be the only option. Since this is talking about features, does it mean to say `Bucket root keys that do not depend on the name of the bucket` instead?

Reply:
The goal of this point to not use the bucket name to derive the key in the default case so that we only need to worry about the server side to perform renames. Specifically, the encrypted paths depend on the bucket key, so if renaming the bucket changes what key would be used, then we would have to re-encrypt all of the paths with the new key.

Why is it important to support deriving a bucket key from some root key and the bucket name in particular? This design supports deriving bucket keys from some root key and some bucket entropy that isn't tied to the name.

Reply:
Oh okay. I misunderstood the point of this doc.

I'm not convinced that the existing bucket situation is problematic. S3 seems to have done just fine without supporting bucket renames at all. If we can support renames without any extra coordination steps, then that sounds great, but I'm not sure how it will work.

Reply:
I agree. This shouldn't require extra coordination steps.